packages/kbn-test/src/functional_test_runner/lib/suite_tracker.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import fs from 'fs';
+import { dirname, relative, resolve } from 'path';
+
+import { REPO_ROOT } from '@kbn/dev-utils';
+
+import { Lifecycle } from './lifecycle';
+
+export interface SuiteInProgress {
+  startTime?: Date;
+  endTime?: Date;
+  success?: boolean;
+}
+
+export interface SuiteWithMetadata {
+  config: string;
+  file: string;
+  tag: string;
+  title: string;
+  startTime: Date;
+  endTime: Date;
+  duration: number;
+  success: boolean;
+  leafSuite: boolean;
+}
+
+const getTestMetadataPath = () => {
+  return process.env.TEST_METADATA_PATH || resolve(REPO_ROOT, 'target', 'test_metadata.json');
+};
+
+export class SuiteTracker {
+  lifecycle: Lifecycle;
+  finishedSuitesByConfig: Record<string, Record<string, SuiteWithMetadata>> = {};
+  inProgressSuites: Map<object, SuiteInProgress> = new Map<object, SuiteInProgress>();
+
+  getTracked(suite: object): SuiteInProgress {
+    if (!this.inProgressSuites.has(suite)) {
+      this.inProgressSuites.set(suite, { success: true } as SuiteInProgress);
```

Comment:
Seems strange to initialize the `SuiteInProgress` object with `success: true`

Reply:
The most straightforward alternative is probably to have success be null/undefined until either the end is reached, or an error occurs, whichever happens first. If the end is reached and success is still not set to false, it would be set to true. I don't have a strong opinion either way

Reply:
I think undefined makes the most sense, since the success state of the suite is not defined until it completes one way or the other. It probably won't ever be an issue but it might be a little more understandable three years from now when someone else cracks open this code and tries to understand what's going on.